contrib/jinzhu/gorm/gorm.go

```diff
@@ -7,17 +7,109 @@
 package gorm
 
 import (
+	"context"
+	"math"
+	"time"
+
 	sqltraced "gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
 
 	"github.com/jinzhu/gorm"
 )
 
+const (
+	gormContextKey       = "dd-trace-go:context"
+	gormConfigKey        = "dd-trace-go:config"
+	gormSpanStartTimeKey = "dd-trace-go:span"
+)
+
 // Open opens a new (traced) database connection. The used dialect must be formerly registered
 // using (gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql).Register.
-func Open(dialect, source string) (*gorm.DB, error) {
-	db, err := sqltraced.Open(dialect, source)
+func Open(dialect, source string, opts ...Option) (*gorm.DB, error) {
+	sqldb, err := sqltraced.Open(dialect, source)
 	if err != nil {
 		return nil, err
 	}
-	return gorm.Open(dialect, db)
+	db, err := gorm.Open(dialect, sqldb)
+	if err != nil {
+		return db, err
+	}
+	return WithCallbacks(db, opts...), err
+}
+
+// WithCallbacks registers callbacks to the gorm.DB for tracing.
+func WithCallbacks(db *gorm.DB, opts ...Option) *gorm.DB {
+	afterFunc := func(operationName string) func(*gorm.Scope) {
+		return func(scope *gorm.Scope) {
+			after(scope, operationName)
+		}
+	}
+
+	cb := db.Callback()
+	cb.Create().Before("gorm:before_create").Register("dd-trace-go:before_create", before)
+	cb.Create().After("gorm:after_create").Register("dd-trace-go:after_create", afterFunc("gorm.create"))
+	cb.Update().Before("gorm:before_update").Register("dd-trace-go:before_update", before)
+	cb.Update().After("gorm:after_update").Register("dd-trace-go:after_update", afterFunc("gorm.update"))
+	cb.Delete().Before("gorm:before_delete").Register("dd-trace-go:before_delete", before)
+	cb.Delete().After("gorm:after_delete").Register("dd-trace-go:after_delete", afterFunc("gorm.delete"))
+	cb.Query().Before("gorm:query").Register("dd-trace-go:before_query", before)
+	cb.Query().After("gorm:after_query").Register("dd-trace-go:after_query", afterFunc("gorm.query"))
+	cb.RowQuery().Before("gorm:row_query").Register("dd-trace-go:before_row_query", before)
+	cb.RowQuery().After("gorm:row_query").Register("dd-trace-go:after_row_query", afterFunc("gorm.row_query"))
+
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	return db.Set(gormConfigKey, cfg)
+}
+
+// WithContext returns a new gorm.DB with the context added
```

Comment:
Because I didn't pay much attention to this whole issue and thread, it took me some time to figure out how this can be used. But maybe it's just me... If so, let me know.

Should we mention here what this is used for? e.g.
```go
// WithContext returns a new gorm.DB with the given context added. The context
// will be used in creating subsequent spans which will be seen as children of any
// span found in ctx.
```
Or something similar 🙂

Reply:
I have an example that was used for E2E tests. Either a snippet from that could be used to explain it, or the package docs can link to the example(s) after they're published.

It gets used like this:
```go
type crudServer struct{
        db *gorm.DB
}

func (c crudServer) Create(ctx context.Context, req *crud.CreateRequest) (*crud.CreateResponse, error) {
        db := gormtrace.WithContext(ctx, c.db)
        t := &Thing{Name: req.GetName()}
        db.Create(t)
        if db.Error != nil {
                return &crud.CreateResponse{Id: 0}, nil
        }
        return &crud.CreateResponse{Id: int32(t.ID)}, nil
}
.. other service methods
```

Reply:
I think something like this should suffice:
```go
// WithContext attaches the specified context to the given db. It may be used
// to link child spans to spans already present in ctx.
```